src/blockchain/rpc.rs

```diff
@@ -197,132 +207,151 @@ impl Blockchain for RpcBlockchain {
+        }
 
-        //TODO supported up to 1_000 txs, should use since_blocks or do paging
-        let list_txs = self
-            .client
-            .list_transactions(None, Some(1_000), None, Some(true))?;
-        let mut list_txs_ids = HashSet::new();
-
-        for tx_result in list_txs.iter().filter(|t| {
-            // list_txs returns all conflicting tx we want to
-            // filter out replaced tx => unconfirmed and not in the mempool
-            t.info.confirmations > 0 || self.client.get_mempool_entry(&t.info.txid).is_ok()
-        }) {
-            let txid = tx_result.info.txid;
-            list_txs_ids.insert(txid);
-            if let Some(mut known_tx) = known_txs.get_mut(&txid) {
-                let confirmation_time =
-                    ConfirmationTime::new(tx_result.info.blockheight, tx_result.info.blocktime);
-                if confirmation_time != known_tx.confirmation_time {
-                    // reorg may change tx height
-                    debug!(
-                        "updating tx({}) confirmation time to: {:?}",
-                        txid, confirmation_time
-                    );
-                    known_tx.confirmation_time = confirmation_time;
-                    db.set_tx(known_tx)?;
-                }
-            } else {
-                //TODO check there is already the raw tx in db?
-                let tx_result = self.client.get_transaction(&txid, Some(true))?;
-                let tx: Transaction = deserialize(&tx_result.hex)?;
-                let mut received = 0u64;
-                let mut sent = 0u64;
-                for output in tx.output.iter() {
-                    if let Ok(Some((kind, index))) =
-                        db.get_path_from_script_pubkey(&output.script_pubkey)
-                    {
-                        if index > *indexes.get(&kind).unwrap() {
-                            indexes.insert(kind, index);
-                        }
-                        received += output.value;
-                    }
+        // list_transactions returns conflicting txs too.
+        // Filter out conflicting: confirmation < 0 and not in mempool.
+        let listed_txs = listed_txs
+            .iter()
+            .filter(|t| {
+                t.info.confirmations > 0 || self.client.get_mempool_entry(&t.info.txid).is_ok()
+            })
+            .collect::<Vec<_>>();
+
+        let batch_update = loop {
+            request = match request {
+                Request::Script(script_req) => {
+                    let scripts = script_req.request();
+
+                    let satisfier = scripts
+                        .map(|sk| {
+                            let related_txs = listed_txs
+                                .iter()
+                                .filter(|tx_result| {
+                                    // Filter out txs related to the script_pubkey
+                                    if let Some(address) = &tx_result.detail.address {
```

Comment:
I don't think we can rely on this to filter txs, what happens for example if we have a tx with 2 owned outputs ?
I am not sure what's the best way to handle this since at this stage we haven't the raw tx, but I think we should operate on raw tx to check the presence of script_pubkey

Reply:
I think in that case it will just match for both the script_pubkeys. `listtransactions()` will give an entry corresponding to each owned addresses. So the `satisfier` will have the same txid for both the scripts. In fact this is happening already in some of the tests, and I guess the underlying db update logic already handles that.

Saying that I do think the current filter logic is inadequate, and because of that the tests are failing. see https://github.com/bitcoindevkit/bdk/pull/466#discussion_r745873363

Reply:
> listtransactions() will give an entry corresponding to each owned addresses.

I didn't know about this, should be ok then